### ISSUE TITLE
Negotiationless codec selection and per-encoding layer codec configuration

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-extensions#126]
## Problem
With the introduction of [scalabilityMode](https://w3c.github.io/webrtc-svc/#dom-rtcrtpencodingparameters-scalabilitymode) (webrtc-svc spec), we now have two competing APIs for configuring the encoder: setCodecPreferences (or SDP munging) and setParameters() with scalabilityMode. Issues:

1. Different codecs have different scalability modes and toggling "simulcast VP8 without SVC" and "non-simulcast VP9 using SVC" not only changes scalability, it also changes the number of active layers. Not being able to set all settings (=codec, scalability mode, active) in a single API call inevitably leads to temporarily sending the wrong thing. Its racy, causes additional key frames and not very ergonomic.

Other issues not specific to SVC:

2. Some apps don't want to re-negotiate (e.g. fixed number of transceiver setups), besides, why do a round-trip to the server if you already know the server supports the desired codec?

3. An app may want to use different codec configurations for different layers e.g. performance vs bitrate tradeoffs.

## Proposal
Configure send codecs with setParameters(). Add "codec" to [RTCRtpEncodingParameters](https://w3c.github.io/webrtc-pc/#rtcrtpencodingparameters) of type [RTCRtpCodecCapability](https://w3c.github.io/webrtc-pc/#dom-rtcrtpcodeccapability) (same type that is used by getCapabilities and setCodecPreferences).

This means that the codec list order in the SDP will not dictate what codec we use for sending. What about backwards-compatibility? Don't worry, we still default to the first codec in the list, but if the codec value is ever changed, that's what we use going forward.

## Pseudo-PR

Update RTCRtpSender initialization steps:
1. Let sender have a [[PreferSelectedCodecs]] internal slot, initialized to false.
2. Let each encoding in sender's [[SendEncodings]] have a [[SelectedCodec]] internal slot representing the currently selected codec for this encoding, initialized to null.

Update the RTCRtpSender.setParameters() steps:
1. For each encoding in the argument to this function, let codec be the value of encodings[i].codec:
1.1. If codec is not one of the values in RTCRtpSender.getCapabilities(kind).codecs, throw an exception and abort these steps.
1.2. If transceiver.mid is not null and codec is not one of the values in [[SendCodecs]], throw an exception and abort these steps. _(If transceiver.mid is null we've not negotiated yet and allow optimistically selecting one.)_
1.3. If codec is not equal to the encoding's [[SelectedCodec]], set the encoding's [[SelectedCodec]] to codec and set [[PreferSelectedCodec]] to true. In parallel, reconfigure the this encoding's encoder to use codec.

Update the SDP negotiation steps, where [[SendCodecs]] is updated and after the number of [[SendEncodings]] have been established:
1. For each encoding in [[SendEncodings]]:
1.1. If [[PreferSelectedCodecs]] is false, or if the encoding's [[SelectedCodec]] is not in the list of [[SendCodecs]], set the encoding's [[SelectedCodec]] to the first element in [[SendCodecs]]. _(Handles default value and backwards-compat for apps relying on setCodecPreferences to change codec.)_
1.2. Reconfigure the encoding's encoder to use [[SelectedCodec]].

**Comment by henbos** [comment]
Nvm I transferred the issue!

**Comment by henbos** [comment]
> 1.2. If transceiver.mid is not null and codec is not one of the values in [[SendCodecs]], throw an exception and abort these steps. (If transceiver.mid is null we've not negotiated yet and allow optimistically selecting one.)

I noticed that [[SendCodecs]] is not set until setting the answer. We should also be able to do setParameters() in-between setRemoteDescription(offer) and createAnswer() for when SRD creates a transceiver and we want to send with it and make sure the codec is as desired.

We could modify this to "If transceiver.currentDirection is not null" to loosen this check.

**Comment by henbos** [comment]
Side-issue: After the complete negotiation, you can check which possible values for codecs exist using getParameters().codecs. However because this only gets updated at the time of setting the answer, you don't know what codecs are valid to pick in-between SRD(offer) and SLD(answer). It would be nice if you could check getParameters().offeredCodecs so that you don't need to parse SDP in order to avoid falling back on the first codec in the list if the one you want to select is not supported by the other endpoint. It would also make sense to throw an exception if at the time of setParameters() we can already deduce that your selection won't work (unlike if you set before negotiation).

**Comment by ssilkin** [comment]
There is a logic in WebRTC which changes encoder on failure (create, init or encode). For example: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/video/video_stream_encoder.cc;drc=acabb3641b924dd0c5198a16883ae8326f35487e;l=1396

It removes the failed codec from the list of negotiated ones: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/media/engine/webrtc_video_engine.cc;l=891-892;drc=2df668b7cbf6c1d0766b6ee0ae8147adc8830f2e;bpv=1;bpt=1. And forces usage of the next negotiated codec by that.

How do solve missing encoding parameters for this case? Can we make it possible providing encoding parameters for multiple codecs?

**Comment by fippo** [comment]
Note that you also can't change the codec sent currently with setParameters. This requires (local) O/A with some SDP manipulation to change the order of codecs in the m= line.

**Comment by ssilkin** [comment]
I just wanted to highlight that there are cases when even if you force a codec via encoding parameters that doesn't guarantee that the codec will actually be used. And if actual codec is different from what we asked for in encoding parameter then we end up with codec & params out-of-sync issue again (which this proposal aims to solve).

**Comment by henbos** [comment]
Got it. Well setParameters() returns a promise, so if it cannot apply the parameters then the right thing to do is to reject that promise. If encoder errors happens later, then we have the same problem is no different from what it is today (you expect one codec, but you get a different one). If this is a real problem I think we could discuss an encoder error callback, but that is a separate issue.

**Comment by ssilkin** [comment]
Can we make it possible to provide encoding parameters for all negotiated codecs at once? That would allow WebRTC  to apply the right parameters whenever codec changes.

**Comment by aboba** [comment]
@ssilkin Yes, it is possible for `setParameters()` and O/A to be out of sync, because the codec constraints change as the negotiation proceeds: 

a. Prior to completion of O/A and determination of the "simulcast envelope", `codec` needs to be one of  `RTCRtpSender.getCapabilities(kind).codecs`.  

b. After completion of O/A and deterrmination of the "simulcast envelope", `codec` needs to be one of the negotiated codecs.

c. `codec` could be a valid value prior to completion of O/A, but invalid afterwards.  For example, mixed codec simulcast (e.g. AV1 at low resolution and H.264 at higher resolutions) could have been set prior to completion of O/A, but O/A negotiated VP8 at all resolutions.  Rather than throwing an exception, it seems like the application needs to use `getParameters()` to discover the settings and then use `setParameters()` to reset them.

**Comment by henbos** [comment]
@Orphis has offered to be the driver of this issue
